### DISCUSSION TITLE
Is there a minimum browser version/feature policy for MQTT.js? (BigInt support issues)

### DISCUSSION THREAD

**Discussion by lorennorman** [mqttjs/MQTT.js#1881] [General]
Hello!

We're running into issues updating from version 4.2.x to 5+ for our legacy browser builds. Technical details follow, but I think it would be most helpful if this library was clear about which browser versions it supports.

Is there a document anywhere detailing the minimum browser requirements for MQTT.js?

Through much experimentation, we've determined they are _at least_:
- Chrome 67
- Edge 79
- Safari (Mobile) 14
- Firefox 68
- Opera 54
- Internet Explorer: unsupported

This appears to be due to:
- MQTT.js bumping a dependency: `readable-stream`
- `readable-stream` is a browser port of a Node built in library
- Node uses `BigInt` constructors and literals
- as of `readable-stream` 4+, it [embeds BigInt literals into its code](https://github.com/nodejs/readable-stream/blob/a2e9aedf4aeee4a5e4d8efcb175edb67e2817eaa/src/errors.js#L339)
- BigInt is not capable of being transpiled or polyfilled for lower browser versions ([Why?](https://github.com/GoogleChromeLabs/jsbi?tab=readme-ov-file#why))

I feel this is worth pointing out because the whole thing is so confusing and it's wasting many developers' time. We get JS compilation issues that we aren't used to ("why does everything else transpile but not this? what code is even using this?", etc.) We get brief responses to issues on various libraries that seem like the maintainer may not fully understand the issue (even the build tool maintainers.)

Love to hear any thoughtful comments on all this!

2 possible recommendations:
- (easy/quick) update the README to mention minimum browser versions and the greater BigInt issue
- (harder) update the project to defend against this: downgrade the dependency, use something else, etc

**Comment by robertsLando** [comment]
Hi @lorennorman and thanks for your findings.

I started maintaining MQTTjs from v5 and apart from completely rewrite it in TS I did lot of effort in order to improve MQTTjs browser support by moving from browserify to esbuild, fixing a tons of issues (**TONS**) and finally adding browser tests (including service workers) using wtr.

I personally use MQTTjs in production in many projects and I never had such compatibility issues and I never worried about supporting old browsers version as I think the most important thing is to support latest. About IE not supported I once heard that: _"Breaking IE is a feature not a bug"_ 😄

Jokes apart, about the 2 recomendations I'm open for both but I don't have time to invest on this so I let you decide if:

- Open a PR to update README.md
- Open a PR to improve browser support

I wish we could get rid of `readable-stream` but unfortunately MQTTjs has been firstly built for NodeJS env and so it's all stream based and streams doesn't exist on browser natively (not yet at least). There is a very nice MQTT library that maybe could work better for you: https://www.npmjs.com/package/u8-mqtt

cc @mcollina

**Reply by lorennorman** [reply]
Thank you @robertsLando! More gratitude:
- thank you for taking over this excellent project and keeping it vital!
- thank you taking the time to answer here, in a straightforward and polite manner
- thank you for inviting PRs and suggesting viable alternatives ("zero dependencies" 🤤)

I love that you've stated these views plainly:

> never worried about supporting old browsers

> most important thing is to support latest

> "Breaking IE is a feature not a bug" 

Copy/paste into the README and you're all set!

FWIW, we hold (basically) the same view and may drop all browsers before 2020 as well. ("Never forget IE 6" 🤪) The wrinkle here was the pain of figuring out what was going on.

Cheers!

**Reply by robertsLando** [reply]
> The wrinkle here was the pain of figuring out what was going on.

I absolutely understand that in some cases this can be a pain to debug, working with browser compatibility is always a pain based on my personal experience.

Let me update README so with browser support informations
